### PR TITLE
43 events gallery of photos

### LINKED ISSUE #43: events: gallery of photos
An event (and maybe a project too) should have a gallery displaying an arbitrary number of photos of the event after it occurred.

We should add the path of these photos to the model and create endpoints capable of storing and deleting those photos. We should also ensure the `GET project/{id}` endpoint includes this gallery.

### PR DESCRIPTION
**Pull request by AvilaAndre**
Closes #43 

Creates the endpoints to manage events' photos.

# Review checklist
-   [ ] Behavior is as expected
-   [ ] Clean, well structured code
-   [ ] Properly documents API changes in the corresponding controller test
-   [ ] Contains enough appropriate tests
-   [ ] Changes are reflected in the Wiki if necessary

### CODE REVIEW

**Comment by MRita443** [comment]
Would it be possible to generalize the gallery to activity? We were discussing this today and think that projects benefit from a gallery too for all the images in their details page

**Comment by AvilaAndre** [comment]
> Would it be possible to generalize the gallery to activity? We were discussing this today and think that projects benefit from a gallery too for all the images in their details page

Of course! I'll get onto it :)

**Comment by MRita443** [comment]
Hi @AvilaAndre are the alterations now ready?

File: src/main/kotlin/pt/up/fe/ni/website/backend/model/Activity.kt [hunk 1]
```diff
@@ -43,6 +43,8 @@ abstract class Activity(
     @field:NotBlank
     open var image: String,
 
+    val gallery: MutableList<String> = mutableListOf(),
```
Comment (MRita443) [inline src/main/kotlin/pt/up/fe/ni/website/backend/model/Activity.kt]:
```suggestion
    @ElementCollection
    open val gallery: MutableList<String> = mutableListOf(),
```

Comment (MRita443) [reply src/main/kotlin/pt/up/fe/ni/website/backend/model/Activity.kt]: To tackle the `Basic attribute type should not be a container.` error

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt [hunk 2]
```diff
@@ -83,4 +84,35 @@ abstract class AbstractActivityService<T : Activity>(
         activity.teamMembers.removeIf { it.id == idAccount }
         return repository.save(activity)
     }
+
+    fun addGalleryPhoto(activityId: Long, image: MultipartFile): Activity {
+        if (!repository.existsById(activityId)) {
+            throw NoSuchElementException(ErrorMessages.eventNotFound(activityId))
```
Comment (MRita443) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt]: Here you are throwing a message about an `Event` instead of an `Activity`. If your intention is to verify if the `Activity` exists, the `getActivityById(activityId)` already does that

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt [hunk 3]
```diff
@@ -83,4 +84,35 @@ abstract class AbstractActivityService<T : Activity>(
...
+
+        activity.gallery.add(imageName)
+
+        return repository.save(activity)
+    }
+
+    fun removeGalleryPhoto(activityId: Long, photoName: String): Activity {
+        if (!repository.existsById(activityId)) {
+            throw NoSuchElementException(ErrorMessages.eventNotFound(activityId))
+        }
```
Comment (MRita443) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt]: Same as above

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/ErrorMessages.kt [hunk 4]
```diff
@@ -44,6 +44,8 @@ object ErrorMessages {
 
     fun roleNotFound(id: Long): String = "role not found with id $id"
 
+    fun photoNotFound(): String = "photo not found"
```
Comment (MRita443) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/ErrorMessages.kt]: I think it would be beneficial to include the image's name here, and perhaps make it more general by using "Image" instead of "Photo"

File: src/main/kotlin/pt/up/fe/ni/website/backend/controller/EventController.kt [hunk 5]
```diff
@@ -74,4 +74,18 @@ class EventController(private val service: EventService) {
         @PathVariable idEvent: Long,
         @PathVariable idAccount: Long
     ) = service.removeTeamMemberById(idEvent, idAccount)
+
+    @PutMapping("/{idEvent}/gallery/addPhoto", consumes = ["multipart/form-data"])
```
Comment (MRita443) [inline src/main/kotlin/pt/up/fe/ni/website/backend/controller/EventController.kt]: @Sirze01 Could you verify the restfulness of these endpoints ;P?

File: src/test/kotlin/pt/up/fe/ni/website/backend/utils/documentation/payloadschemas/model/PayloadActivity.kt [hunk 6]
```diff
@@ -89,6 +89,13 @@ class PayloadActivity {
                     "Description of the event",
                     JsonFieldType.STRING,
                     optional = true
+                ),
+                DocumentedJSONField(
+                    "gallery[]",
+                    "Array of photos associated with the activity",
+                    JsonFieldType.ARRAY,
+                    isInRequest = false,
```
Comment (MRita443) [inline src/test/kotlin/pt/up/fe/ni/website/backend/utils/documentation/payloadschemas/model/PayloadActivity.kt:102]: Given how the image of each `Activity` is being documented thus far, I'd leave leave this line out and specify that what is being saved is the path to these images

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt [hunk 7]
```diff
@@ -83,4 +84,35 @@ abstract class AbstractActivityService<T : Activity>(
...
+
+    fun addGalleryPhoto(activityId: Long, image: MultipartFile): Activity {
+        if (!repository.existsById(activityId)) {
+            throw NoSuchElementException(ErrorMessages.eventNotFound(activityId))
+        }
+
+        val activity = getActivityById(activityId)
+
+        val fileName = fileUploader.buildFileName(image, activity.title)
+        val imageName = fileUploader.uploadImage("gallery", fileName, image.bytes)
```
Comment (MRita443) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt]: Here it would be best to prefix this path with the `IMAGE_FOLDER` of each `Activity`, in order to separate them , no? Either that or just that same exact folder, what do you think?

**Review by MRita443 — CHANGES_REQUESTED** [review]
Few suggestions so far, will look at the tests after these changes :)

File: src/main/kotlin/pt/up/fe/ni/website/backend/model/Activity.kt (hunk 1 above)
Comment (AvilaAndre) [reply src/main/kotlin/pt/up/fe/ni/website/backend/model/Activity.kt]: In order to use @ElementCollection the fetch type must be changed to Eager because by default it is Lazy, being Lazy the value must be obtained through a query or else it throws an exception, that's why my tests were failing en masse :p

File: src/main/kotlin/pt/up/fe/ni/website/backend/controller/EventController.kt [hunk 8]
```diff
@@ -74,4 +74,18 @@ class EventController(private val service: EventService) {
...
+        @RequestParam
+        @ValidImage
+        image: MultipartFile
+    ) = service.addGalleryImage(idEvent, image)
+
+    @PutMapping("/{idEvent}/gallery/removeImage")
+    fun removeGalleryImage(
+        @PathVariable idEvent: Long,
+        @RequestPart imageUrl: String
+    ) = service.removeGalleryImage(idEvent, imageUrl)
```
Comment (BrunoRosendo) [inline src/main/kotlin/pt/up/fe/ni/website/backend/controller/EventController.kt]: As rita pointed out, this should be changed according to #180

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/ErrorMessages.kt [hunk 9]
```diff
@@ -44,6 +44,8 @@ object ErrorMessages {
 
     fun roleNotFound(id: Long): String = "role not found with id $id"
 
+    fun imageNotFound(imageName: String): String = "image not found with name $imageName"
```
Comment (BrunoRosendo) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/ErrorMessages.kt]: I would personally generalize this to `fileNotFound`, so we can use it for other files

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt [hunk 10]
```diff
@@ -74,13 +79,40 @@ abstract class AbstractActivityService<T : Activity>(
...
+    fun addGalleryImage(activityId: Long, image: MultipartFile): Activity {
+        val activity = getActivityById(activityId)
+
+        var imageFolder = "activities"
+
+        if (activity is Event) {
+            imageFolder = EventService.IMAGE_FOLDER
+        } else if (activity is Project) {
+            imageFolder = ProjectService.IMAGE_FOLDER
+        }
```
Comment (BrunoRosendo) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt]: I think this is a big violation of OOP and we can improve this. My suggestion is to a method like this in the abstract activity service:

```suggestion
fun addGalleryImage(activityId: Long, image: MultipartFile, imageFolder: String): Activity {
        val activity = getActivityById(activityId)
        
        val fileName = fileUploader.buildFileName(image, activity.title)
        val imageName = fileUploader.uploadImage(imageFolder + "/gallery", fileName, image.bytes)

        activity.gallery.add(imageName)

        return repository.save(activity)
    }
```

And then add new methods to the project/events services where the name is chosen. This would maintain the logic and decouple the inheritance logic from the abstract service

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt [hunk 11]
```diff
@@ -74,13 +79,40 @@ abstract class AbstractActivityService<T : Activity>(
...
+
+        val fileName = fileUploader.buildFileName(image, activity.title)
+        val imageName = fileUploader.uploadImage(imageFolder + "/gallery", fileName, image.bytes)
+
+        activity.gallery.add(imageName)
+
+        return repository.save(activity)
+    }
+
+    fun removeGalleryImage(activityId: Long, imageName: String): Activity {
```
Comment (BrunoRosendo) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt:97]: We should also erase the photo from storage!

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 12]
```diff
@@ -271,6 +274,89 @@ internal class EventControllerTest @Autowired constructor(
         }
     }
 
+    @NestedTest
+    @DisplayName("GET events/category/{category}")
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:278]: Was this not tested before? I'm a bit confused since it seems out of scope for this PR

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 13]
```diff
@@ -725,6 +812,169 @@ internal class EventControllerTest @Autowired constructor(
...
+                .addFile("image", contentType = MediaType.IMAGE_JPEG_VALUE)
+                .perform()
+                .andExpectAll(
+                    status().isOk,
+                    content().contentType(MediaType.APPLICATION_JSON),
+                    jsonPath("$.title").value(testEvent.title),
+                    jsonPath("$.description").value(testEvent.description),
+                    jsonPath("$.teamMembers.length()").value(testEvent.teamMembers.size),
+                    jsonPath("$.gallery.length()").value(1),
+                    jsonPath("$.gallery[0]").value(expectedImagePath),
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:857]: Have you (manually) tested if you can access this path on our machine (e.g. open in your browser)? I assume this was already tested for other images but I don't fully remember

**Review by BrunoRosendo — CHANGES_REQUESTED** [review]
good job, this is looking good! I got a few suggestions

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt (hunk 12 above)
Comment (AvilaAndre) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:278]: I honestly don't remember this part, my guess is that I updated the branch and needed to modify it to include the gallery.

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt (hunk 10 above)
Comment (AvilaAndre) [reply src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt]: Changed the highlighted code as proposed, while keeping the companion objects.

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt (hunk 13 above)
Comment (AvilaAndre) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:857]: It exists on the filesystem but the url does not work (I don't know if I am missing something because the port is 3000 and there is nothing there).

Comment (MRita443) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:857]: @BrunoRosendo

Comment (BrunoRosendo) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:857]: Please check #71 again, I'm pretty sure @DoStini already had the logic for this. There is a folder where Spring will save public files (see https://github.com/NIAEFEUP/website-niaefeup-backend/wiki/File-Upload#static-file-uploader)

I suspect the servePath is not being used correctly

Comment (AvilaAndre) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:857]: Images are being stored locally in the `build/resources/main/static` folder, Spring Boot automatically serves the `src/main/resources/static` folder, however, due to the `endpointNotFound` present on the `ErrorController` (`src/main/kotlin/pt/up/fe/ni/website/backend/controller/ErrorController`) having a `RequestMapping("/**")`, it overrides the default behaviour and responds with an error 404 "invalid endpoint". Altough images are not being served, they also are not in the correct place for it, as they would not be served in the build directory.
To test this behaviour, remove the two `*` in the `RequestMapping` mentioned (leaving only `/`) and create an about.html with a `Hello World` string inside, you will see that it exists in `http://localhost:8080/about.html`.

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/upload/CloudinaryFileUploader.kt [hunk 14]
```diff
@@ -20,4 +20,11 @@ class CloudinaryFileUploader(private val basePath: String, private val cloudinar
 
         return result["url"]?.toString() ?: ""
     }
+
+    override fun deleteImage(fileName: String) {
+        cloudinary.uploader().destroy(
+            fileName,
```
Comment (MRita443) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/upload/CloudinaryFileUploader.kt:27]: Shouldn't the file here be removed based on the the same path as in the upload, `"$basePath/$folder/$fileName"`?

File: src/main/resources/application.properties [hunk 15]
```diff
@@ -32,9 +32,14 @@ upload.provider=static
...
 upload.cloudinary-url=GET_YOURS_AT_CLOUDINARY_DASHBOARD
 # Folder in which files will be stored
-upload.static-path=classpath:static
+upload.static-path=file:static
 # URL that will serve static content
-upload.static-serve=http://localhost:3000/static
+upload.static-serve=http://localhost:8080/static
+
+# Files are served in the following path
+spring.mvc.static-path-pattern=/static/**
```
Comment (MRita443) [inline src/main/resources/application.properties:40]: What does this do?

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 16]
```diff
@@ -865,7 +865,7 @@ internal class EventControllerTest @Autowired constructor(
         fun `should fail if event does not exist`() {
             val unexistentID = 5
```
Comment (MRita443) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]:
```suggestion
            val nonexistentID = 5
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 17]
```diff
@@ -946,8 +947,8 @@ internal class EventControllerTest @Autowired constructor(
...
```
Comment (MRita443) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]:
```suggestion
            val nonexistentID = 5
```

File: src/test/kotlin/pt/up/fe/ni/website/backend/utils/mockmvc/MockMvcMultipartBuilder.kt [hunk 18]
```diff
@@ -49,6 +49,16 @@ class MockMvcMultipartBuilder(
...
+    fun asDeleteMethod(): MockMvcMultipartBuilder {
+        multipart.with(
+            {
+                it.method = "DELETE"
+                it
+            }
+        )
+        return this
+    }
+
```
Comment (MRita443) [inline src/test/kotlin/pt/up/fe/ni/website/backend/utils/mockmvc/MockMvcMultipartBuilder.kt]:
```suggestion
   fun asDeleteMethod(): MockMvcMultipartBuilder {
        multipart.with {
            it.method = "DELETE"
            it
        }
        return this
    }
```

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt [hunk 19]
```diff
@@ -85,17 +83,9 @@ abstract class AbstractActivityService<T : Activity>(
...
-        var imageFolder = "activities"
-
-        if (activity is Event) {
-            imageFolder = EventService.IMAGE_FOLDER
-        } else if (activity is Project) {
-            imageFolder = ProjectService.IMAGE_FOLDER
-        }
-
         val fileName = fileUploader.buildFileName(image, activity.title)
         val imageName = fileUploader.uploadImage(imageFolder + "/gallery", fileName, image.bytes)
```
Comment (MRita443) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/activity/AbstractActivityService.kt]:
```suggestion
        val imageName = fileUploader.uploadImage("$imageFolder/gallery", fileName, image.bytes)
```

**Review by MRita443 — CHANGES_REQUESTED** [review]
Some things:
- Tests are missing for `Project`
- Tests are missing the documentation functions

Rest is mostly questions and things my IDE nagged me about :)

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/upload/CloudinaryFileUploader.kt (hunk 14 above)
Comment (AvilaAndre) [reply src/main/kotlin/pt/up/fe/ni/website/backend/service/upload/CloudinaryFileUploader.kt:27]: I'll add a comment above this line, but basically, it is called `fileName` due to the overriding the method in the supertype `FileUploader`, however, this is actually the `public_id` of the resource in Cloudinary, with will be the path. You can see this `public_id` in the method `uploadImage` above.

File: src/main/resources/application.properties (hunk 15 above)
Comment (AvilaAndre) [reply src/main/resources/application.properties:40]: The first line I changed the port to be the same as where the server is running, because that is where the files will be located if we are using the local filesystem, the second line tells Spring where the static files to be served are.

**Comment by MRita443** [comment]
I was just testing this out on Postman and noticed a couple more things
image

The image of the event itself has an incorrect port, perhaps we should fix that in this PR too.
Also, I think maybe using the slug would be better than the title for the filename

**Comment by MRita443** [comment]
Also, regarding the documentation, I don't think it'll be possible for now because #144 is still open, my bad!

**Comment by AvilaAndre** [comment]
> I was just testing this out on Postman and noticed a couple more things image
> 
> The image of the event itself has an incorrect port, perhaps we should fix that in this PR too. Also, I think maybe using the slug would be better than the title for the filename

Did that event already exist on your machine?
`3000` is the port that was hardcoded before. I already replaced that with 8080.

**Comment by AvilaAndre** [comment]
> Also, regarding the documentation, I don't think it'll be possible for now because #144 is still open, my bad!

I already added some, the multipart is missing, maybe we just add that after #144 is finished.

Update: Already figured out why I can't use it.

**Comment by MRita443** [comment]
> > I was just testing this out on Postman and noticed a couple more things image
> > The image of the event itself has an incorrect port, perhaps we should fix that in this PR too. Also, I think maybe using the slug would be better than the title for the filename
> 
> Did that event already exist on your machine? `3000` is the port that was hardcoded before. I already replaced that with 8080.

Yep you're right! Created another one and it's correct

**Review by MRita443 — APPROVED** [review]
Great job! This ended up being a bit more than expected but you handled it well, good eye catching all the slug mentions too :)

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 20]
```diff
@@ -725,6 +812,195 @@ internal class EventControllerTest @Autowired constructor(
...
+                    jsonPath("$.image").value(testEvent.image)
+                )
+                /*
+                .andDocument(
+                    documentation,
+                    "Adds an image to the gallery of the selected event",
+                    "This endpoint adds an image to the gallery of an event, returning the event's updated data.",
+                    urlParameters = parameters
+                )
+                */
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:873]: is this not fixed yet? :'(

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 21]
```diff
@@ -725,6 +812,195 @@ internal class EventControllerTest @Autowired constructor(
...
+                    jsonPath("$.image").value(testEvent.image)
+                )
+                /*
+                .andDocument(
+                    documentation,
+                    "Removes an image from the gallery of a selected event",
+                    "This endpoint removes an image to the gallery of an event, returning the event's updated data.",
+                    urlParameters = parameters
+                )
+                */
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:967]: same, others below

**Review by BrunoRosendo — APPROVED** [review]
Great job! :D I can't believe this is finally getting merged 😄

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt (hunk 20 above)
Comment (MRita443) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:873]: No, the `springdoc-openapi` doesn't support this yet :(
